internal/core/service/interfaces.go

```diff
@@ -0,0 +1,49 @@
+package service
+
+//go:generate mockgen -source=$GOFILE -destination=mocks/services.go
+
+import (
+	"context"
+	"github.com/zhuravlev-pe/course-watch/internal/core/domain"
+	"github.com/zhuravlev-pe/course-watch/internal/core/dto"
+)
+
+type UsersRepository interface {
+	GetById(ctx context.Context, id string) (*domain.User, error)
+	Insert(ctx context.Context, user *domain.User) error
+	Update(ctx context.Context, id string, input *dto.UpdateUserInfoInput) error
+	GetByEmail(ctx context.Context, email string) (*domain.User, error)
+}
+
+type CoursesRepository interface {
+	GetById(ctx context.Context, id string) (*domain.Course, error)
+	Insert(ctx context.Context, course *domain.Course) error
+}
+
```

Comment:
If I understood "ports and adapters" correctly, all ports belong to core ("hexagon" or "application") and adapters ("infrastructure") are using usecases as dependencies (driver adapters) or are used by them (driven adapters). But we definitely should ask Nicolas.

Reply:
Yes, dependencies should work as you describe, but the key part for Go is that the interfaces are declared by the consumer, not by the implementation. Moving the interface definition does not change the relationship. It is only a mechanism for getting rid of direct dependencies (import references) between packages which is critical for Go due to circular dependency problem

Reply:
Well, I get your point, but every example, including two, you provided, have had an interface of a service declared together with an implementation, e.g.:

**Driving side**
The driving actors are those that start the interaction with the application by initiating a query. For example, it can be the mobile application interface or user interface code of a web app. The user input passed into the UI is taken by the adapter and sent to the core through the port. _Both the port (interface) and the implementation of the interface will be inside the core/hexagon._

Reply:
Yes, but the examples I've provided are in languages with paradigm different from Go. They don't have type-safe duck typing. That is why they have to define interfaces alongside the code which implement them (well, you _can_ have duck typing similar to Go in TS, but it is not idiomatic to do so there, as far as I know; C# does not have duck typing except for `dynamic` stuff, which is extremely ugly not type safe). 

I could not find good examples of Go code with hexagonal architecture, so it is up to us to figure out how to deal with Go specifics

According to Nicolas we should try to define interfaces in the packages where they are consumed, just like what you did with the repository interfaces. That is why I suggest to do so with the service interfaces and, once this is done, we could ask for more expert opinion from him.

Reply:
As you have pointed out, this pattern is defined for languages that are object oriented. The concept that we should focus on is that our bussines logic is the core of the project and that it should not have any dependency. That means the bussines logic defines every resource that it needs and then we make the "tools" we use (a repository, a client to another service, etc) be compliant with the needs of the bussines logic. In the future, for example, if we decide that we don't want to have our data in a postgres database and we wantto get that from another service by using a gRPC channel, then we just write our client to be compliang with the interface the bussines logic has defined. We should also define the models or dtos in our core and use that structures to communicate with our bussines logic.

I really like that you are having this discussion to try and get it right, but don't get tangled up on the definitions. What is important is to be consistent with the decisions that are being made, like if we have a file called dtos for the structures we should use that naming in every package that we need, or if we call an identifier ID, then we should call identifiers ID everywhere.